core/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -261,6 +261,9 @@ private DocumentMapper merge(DocumentMapper mapper, boolean updateAllTypes) {
             if (mapper.type().length() == 0) {
                 throw new InvalidTypeNameException("mapping type name is empty");
             }
+            if (Version.indexCreated(indexSettings).onOrAfter(Version.V_2_0_0_beta1) && mapper.type().length() > 255) {
```

Comment:
can you add a test for this? You can special the version created in the index settings ;)